src/plugins/scopes/ScopesApi.js

```diff
@@ -0,0 +1,474 @@
+            node,
+            parentScope: this._getParentScopeFor(node),
+            isClassScope: true
+        }));
+    }
+
+    _addClassDeclaration(node: ClassDeclaration) {
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getParentScopeFor(node),
+            isClassScope: true
+        }));
+    }
+
+    _addFunctionExpression(node: FunctionExpression) {
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getParentScopeFor(node),
+            isFunctionScope: true
+        }));
+    }
+
+    _addFunctionDeclaration(node: FunctionDeclaration) {
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getParentScopeFor(node),
+            isFunctionScope: true
+        }));
+    }
+
+    _addArrowFunctionExpression(node: ArrowFunctionExpression) {
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getParentScopeFor(node),
+            isFunctionScope: true,
+            isArrowFunctionScope: true
+        }));
+    }
+
+    _addScopedBlock(node: Node) {
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getParentScopeFor(node)
+        }));
+    }
+
+    _addBlockStatement(node: BlockStatement) {
+        let parentElement = node.parentElement;
+        if (
+            parentElement &&
+            (
+                parentElement.type === 'ForStatement' ||
+                parentElement.type === 'ForInStatement' ||
+                parentElement.type === 'ForOfStatement' ||
+                parentElement.type === 'CatchClause' ||
+                parentElement.type === 'ArrowFunctionExpression' ||
+                parentElement.type === 'FunctionExpression' ||
+                parentElement.type === 'FunctionDeclaration'
+            )
+        ) {
+            return;
+        }
+        this._scopesMap.set(node, new Scope({
+            node,
+            parentScope: this._getScopeFor(parentElement)
+        }));
+    }
+
+    _addJSXIdentifier(node: JSXIdentifier) {
+        let name = node.name;
+        let scope = this._getScopeFor(node);
+        let parentElement = node.parentElement;
+        if (scope && parentElement) {
+            if (parentElement instanceof JSXAttribute) {
+                if (node === parentElement.name) {
+                    return;
+                }
+            }
+            if (parentElement instanceof JSXMemberExpression) {
+                if (node === parentElement.property) {
+                    return;
+                }
+            }
+            if (parentElement instanceof JSXNamespacedName) {
+                if (node === parentElement.name) {
+                    return;
+                }
+            }
+            scope._addReference({node, name, read: true, write: false});
+        }
+    }
+
+    _addIdentifier(node: Identifier) {
```

Comment:
I suppose we really need coverage tools implemented, otherwise it's pretty unclear